### ISSUE TITLE
Summary file races

### ISSUE THREAD

**Issue by dbnicholson** [ostreedev/ostree#487]
Originally from https://github.com/flatpak/flatpak/issues/149.

There appear to be multiple races with the summary files.

1. The previous summary signature file is not removed until the new summary is regenerated. A client could get the old signature and new summary in this window.

2. The summary signature can't be generated until after the new summary file is in place. In addition to the race from 1, if the client has `gpg-verify-summary=true`, then there's a window where no signature exists.

I think there should be a `ostree_repo_regenerate_summary` variant that also takes GPG keys so that ostree can generate the signature and summary in one pass.

To make the race smaller, this new variant would generate the signature and summary in the temporary directory and rename them. However, that's still racy on the server side, and it doesn't fix the race on the client side where the files may be regenerated between the requests for the summary and signature file.

Here's 2 thoughts about making this really atomic:

> The other idea I had to go full atomic would be a summaries directory with a current symlink that can be atomically swapped sure of like the bootloader directory in ostree.

> So, even better would be to have an atomically replaced summary.sig, which contained a metadata field with the unique name of the summary file (we could keep the "summary" name a symlink to the latest summary for backwards compat). Then we could keep old summary files around for a while. This would fix clients racing with the summary being replaced.

CC @alexlarsson

**Comment by dbnicholson** [comment]
#490 does the simple thing of creating the summary file and signing it in a temporary directory and then renaming them. I think the added API is necessary whichever direction this goes, and the implementation could be changed later to become fully atomic.

**Comment by ramcq** [comment]
Could we add support for parsing summary files with inline signatures and just doing it that way...? Or add a summary.insig or such, which we check for first?

**Comment by cgwalters** [comment]
To clarify the status of this PR...I'm not sure it's worth adding API just to *narrow* a race.

This was discussed some on list:
https://mail.gnome.org/archives/ostree-list/2016-October/msg00002.html

I'm pushing basically to fetch the summary file over pinned TLS and not GPG sign it.

We could however investigate doing inline signatures on the summary - this is basically [what Debian did](http://www.chiark.greenend.org.uk/~cjwatson/blog/no-more-hash-sum-mismatch-errors.html).  I'm not opposed to that, but again I think TLS is basically a better match for this, since it helps provide freshness guarantees like Notary does.

**Comment by dbnicholson** [comment]
Hasn't that ship sailed? The signed summary file is part of the ostree repo format for better or worse. I can't see how support for it could be removed in any sane way. And if it's part of ostree, you might as well do it right. Furthermore, since flatpak is requiring summary GPG verification and is now getting wide adoption, it would be a long time before clients stopped requesting summary.sig.

Requiring a separate TLS equipped metadata server also means it's a) much harder to setup an ostree repo, and b) not possible to securely publish an ostree repo on an insecure link.

**Comment by cgwalters** [comment]
I have no plans to remove the signed summary support, but I'm just saying my preference is not to extend it more.

I don't think flatpak *requires* GPG signed summaries - why would it?  I can't see anything in the code or design that does.  (However, the gnome/gnome-apps do indeed use signed summaries today)

Now that Let's Encrypt exists, there's really no excuse for not using TLS.   Whether you do *pinned* TLS is another thing of course.

Define "securely"?  If your commits are signed, you have basic integrity protection.

Philip's work in https://github.com/ostreedev/ostree/pull/782 I believe basically requires *unsigned* summaries since clients may only have partial set of refs from upstream.

**Comment by dbnicholson** [comment]
Sorry, flatpak doesn't require signed summaries. But it sets up the remotes that way by default, so there are likely already tons of remote configurations in the wild where this is the case. Unless you can convince all those people to manually change their remote configurations, you can effectively say the flatpak requires signed summaries for the foreseeable future.

For "securely", I think it's been discussed many times why you would want to sign the summary. Otherwise, why does the infrastructure for signing the summary exist at all? Do you sign the summary for atomic?

#782 is a special case limited by the fact that you're intentionally trying to use an arbitrary, potentially untrusted remote. I definitely tried to think of a way to maintain the chain of trust for the summary when I was reviewing the original work in eos-updater.

**Comment by cgwalters** [comment]
For Fedora/CentOS we don't currently sign the summary, no.  We also don't yet do tls pinning there, though [I am trying](https://lists.fedoraproject.org/archives/list/infrastructure@lists.fedoraproject.org/message/IHRURND4CYMCALBGRSKYUJZEATNXXQEA/) but it blocks a bit on [this issue](https://github.com/ostreedev/ostree/issues/707). For Red Hat Enterprise Linux Atomic Host, we set up TLS pinning by default.

**Comment by alexlarsson** [comment]
Flatpak by default sets up signed summaries, and recently added a feature that relies on it. The summary can have a metadata field that gives a new URL and/or GPG key, allowing a signed summary to transfer the repo to a new location. Of course, it may make sense to move thie signature on this data to just be on the metadata, to allow unsigned summaries in the future.

Other nice things about signed summaries are that it allows mirroring of the summary file, whereas if you rely on TLS for trust it has to be on the main machine. Additionally current ostree either uses all http or all https, and https summary may be ok, but performance wise it seems bad to encrypt all the content.

**Comment by cgwalters** [comment]
Ah...you're talking about https://mail.gnome.org/archives/ostree-list/2017-May/msg00007.html and...hm, lots of flatpak stuff going on here, like https://github.com/flatpak/flatpak/commit/21778f1075e5ea834aeed4469cbfea723ef9ef8c but the main commit appears to be https://github.com/flatpak/flatpak/commit/7a4c82529ec42c961fce341adcc0d8ec9687058a ?

But actually I'm confused since that seems to set a property in the repo config...or I guess you're copying into the summary?

Bear in mind that with the `contenturl` support there's not a "main server" notion necessary anymore.

Anyways this is probably a discussion for the list.

**Comment by alexlarsson** [comment]
Yeah, the new URL is stored in the config and copied into the summary each time it is recreated. Maybe a better approach is to store it signed in the config and copy url + signature into the summary each time.

**Comment by cgwalters** [comment]
I appreciate you get a lot of velocity out of pushing things straight to flatpak git master, but I think for things like this it'd be good to submit a PR and tag people for review.  I'd certainly like to be involved in this particular one as the redirection implementation could be useful for the ostree-host case, and certainly has implications for other ostree-consuming tools (for example [Pulp](https://github.com/pulp/pulp_ostree).

**Comment by cgwalters** [comment]
But anyways let's discuss this one on list, I'll follow up to the thread.

**Comment by pwithnall** [comment]
For the record, the approach I’m looking at [here on the list](https://mail.gnome.org/archives/ostree-list/2017-May/msg00027.html) would solve this problem by deprecating the `summary.sig` and moving the signatures inline into the `summary` and `.commit` files (which area ideas previously discussed in this PR).

Let’s keep discussion on the list; I just wanted to add the cross-reference here.

**Comment by pwithnall** [comment]
An update: we decided to go with an `ostree-metadata` ref which contains the (signed) repository metadata, and to drop the `summary.sig` file (aside from for backwards compatibility), so the `summary` file is now just an unsigned cache map of ref name → checksum. Each commit has signed bindings back to the refs which are allowed to point to it.

We’re using this in EOS now, and it solves the races (aside from for backwards compatibility): a new `ostree-metadata` ref is generated and pushed, and then the atomic switchover from old (refs + metadata + signatures) to new (refs + metadata + signatures) happens by updating the `summary` file to change all the checksums the refs point to. `summary.sig` is still updated, but only for backwards compatibility, so we can ignore that race.

Shall we close this, or do you want to keep it open until the `ostree-metadata` stuff has an API in libostree and is no longer hidden behind `--enable-experimental-api`, @cgwalters?

**Comment by dbnicholson** [comment]
Was thinking about this again in the context of  https://github.com/flatpak/flatpak/pull/3167 and https://lists.freedesktop.org/archives/flatpak/2019-October/001784.html. In particular, the `ostree-metadata` commit does not (as far as I can tell) contain the ref mapping when generated by `ostree summary -u` or `flatpak build-update-repo`.

I think it would be nice if `ostree_repo_regenerate_summary` would just generate the `ostree-metadata` commit if the repo had a collection ID set. Currently both tools just create an empty commit with the additional summary metadata included. That seems simple enough to just include in the summary generation API. Then both the summary and the `ostree-metadata` commit could be changed in one place to include the ref mapping as constructed during the summary generation.

The connection to this issue is that you ideally want to add the signature to the `ostree-metadata` commit in a single transaction. Then the proposed `ostree_repo_regenerate_summary_ext` would do just that since it also is provided GPG keys. I still think the part about narrowing the summary/summary.sig race by using a tempdir and renaming is a nice to have, but not the main point.

Thoughts?

**Comment by dbnicholson** [comment]
Or maybe since this requires new API anyways to provide GPG keys while updating, add `ostree_repo_regenerate_metadata` which updates the summary and `ostree-metadata` commit at the same time while signing them. This would replace the `ostree_repo_regenerate_summary_ext` API suggested in #490.
